src/logcollector/read_journald.c

```diff
@@ -0,0 +1,64 @@
+#include "shared.h"
+#include "logcollector.h"
+#include <systemd/sd-journal.h>
+
+void prime_sd_journal(sd_journal **jrn) {
+  int ret;
+  ret = sd_journal_open(&jrn, SD_JOURNAL_LOCAL_ONLY);
+  ret = sd_journal_get_fd(jrn);
+  ret = sd_journal_seek_tail(jrn);
+  // prime sd_journal_next before the reader loop
+  ret = sd_journal_previous(jrn);
+}
+
+void sd_read_journal(char *journal) {
+  sd_journal *jrn;
+  int ret;
+  const char *data;
+  size_t len;
+  struct timeval tv;
+  uint64_t curr_timestamp;
+  time_t nowtime;
+  struct tm *nowtm;
+  char tmbuf[64], final_msg[OS_MAXSTR];
+
+  prime_sd_journal(&jrn);
+  ret = sd_journal_next(j);
+  // Anything negative is an error
+  while (ret >= 0) {
+    // No messages, wait for data
+    if (ret == 0) {
+      while ((sd_journal_wait(jrn, (uint64_t)-1)) == SD_JOURNAL_NOP) {
+        sleep(1);
+      }
+    } else {
+      // Read data
+      ret = sd_journal_get_data(jrn, "MESSAGE", (const void **)&data, &len);
```

Comment:
`MESSAGE` doesn't include the hostname. Not sure if the log passed into the queue later matches what OSSEC is expecting. When I played with this stuff previously, I also grabbed `_HOSTNAME` and `SYSLOG_IDENTIFIER`

Reply:
Thank you, very useful insight, will play around with these - if you get a chance to do a compile/run test, would love to get feedback on which ones go where.

Reply:
Both _HOSTNAME and SYSLOG_IDENTIFIER are now in there for prodding and reorg, but i think i have all the data requested...